cmd/query/app/querysvc/query_service_v2.go

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package querysvc
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/cmd/query/app/querysvc/adjuster"
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/pkg/iter"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+	"github.com/jaegertracing/jaeger/storage_v2/depstore"
+	"github.com/jaegertracing/jaeger/storage_v2/tracestore"
+)
+
+var errNoArchiveSpanStorage = errors.New("archive span storage was not configured")
+
+const (
+	defaultMaxClockSkewAdjust = time.Second
+)
+
+// QueryServiceOptions has optional members of QueryService
+type QueryServiceOptionsV2 struct {
+	ArchiveTraceReader tracestore.Reader
+	ArchiveTraceWriter tracestore.Writer
+	Adjuster           adjuster.Adjuster
+}
+
+// StorageCapabilities is a feature flag for query service
+type StorageCapabilitiesV2 struct {
+	ArchiveStorage bool `json:"archiveStorage"`
+	// TODO: Maybe add metrics Storage here
+	// SupportRegex     bool
+	// SupportTagFilter bool
+}
+
+// QueryService contains span utils required by the query-service.
+type QueryServiceV2 struct {
+	traceReader      tracestore.Reader
+	dependencyReader depstore.Reader
+	options          QueryServiceOptionsV2
+}
+
+// NewQueryService returns a new QueryService.
+func NewQueryServiceV2(
+	traceReader tracestore.Reader,
+	dependencyReader depstore.Reader,
+	options QueryServiceOptionsV2,
+) *QueryServiceV2 {
+	qsvc := &QueryServiceV2{
+		traceReader:      traceReader,
+		dependencyReader: dependencyReader,
+		options:          options,
+	}
+
+	if qsvc.options.Adjuster == nil {
+		qsvc.options.Adjuster = adjuster.Sequence(adjuster.StandardAdjusters(defaultMaxClockSkewAdjust)...)
+	}
+	return qsvc
+}
+
+// GetTrace is the queryService implementation of tracestore.Reader.GetTrace
+func (qs QueryServiceV2) GetTraces(ctx context.Context, traceIDs ...tracestore.GetTraceParams) iter.Seq2[[]ptrace.Traces, error] {
+	traceIter := qs.traceReader.GetTraces(ctx, traceIDs...)
+	_, err := iter.FlattenWithErrors(traceIter)
```

Comment:
The whole idea of TraceReader returning Seq is to support streaming of the data. By using flatten we would force full data load into query service before it can return anything, even though such aggregation is not required for the things that the query service needs to do. We can process incoming chunks incrementally (see the data contract of the reader) and return them back incrementally as well (which is why we continue with Seq output pattern).

Aren't we supposed to call the adjuster somewhere?

Lastly, `spanstore.ErrTraceNotFound` is no longer used in v2 reader. This makes it slightly more difficult to support the archive lookup. I think we will need to keep a set of IDs returned from the main reader, and once its iterator is finished and not all requested IDs are returned we need to issue a secondary smaller query to archive reader.

Reply:
> Aren't we supposed to call the adjuster somewhere?

The adjuster is called in the [HTTP handler](https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/http_handler.go#L371).

Reply:
Hm, I thought query was taking care of that. Then what does query do? Is its only function to dispatch to archive reader? And why does it have adjuster in the options?

Reply:
@yurishkuro The query service is initialized with the adjuster that is passed in. The `MaxClockSkewAdjust` is a configuration option for the query app (https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/flags.go#L75). To your former point however, the query service just seems to be a wrapper around the reader/writer with archive capabilities.

Reply:
it's ok, we can keep it. It has two extra Archive() and Adjust() functions compared to just the storage reader/writer.

Reply:
Adjust needs to work on Seq as well, both as input and output.

Reply:
@yurishkuro Can you take another pass at the changes and let me know if it makes sense now. I can start writing tests if this looks in the right direction.